specs-go/chain/chainid.go

```diff
@@ -0,0 +1,46 @@
+package chain
+
+import "github.com/docker/distribution/digest"
+
+// ChainID takes a slice of digests and returns the ChainID corresponding to
+// the last entry. Typically, these are a list of layer DiffIDs, with the
+// result providing the ChainID identifying the result of sequential
+// application of the preceeding layers.
+func ChainID(dgsts []digest.Digest) digest.Digest {
+	chainIDs := make([]digest.Digest, len(dgsts))
+	copy(chainIDs, dgsts)
+	ChainIDs(chainIDs)
+
+	if len(chainIDs) == 0 {
+		return ""
```

Comment:
This is a reference implementation for the [definiton][1], so I think we do need an “undefined behavior” error for the undefined case of `ChainID([])`.  Alternatively, we could define the ChainID for that case to be the empty string, but I don't see the point since we [always require at least one layer][2] (with #407 still in flight).

It would have been nice if the chainID algorithm was:

    ChainID([L₀, …, Lₙ₋₁]) = { SHA256hex(''), when n == 0
                             { SHA256hex(ChainID([L₀, …, Lₙ₋₂]) + " " + DiffID(Lₙ₋₁)), when n > 0

to cleanly cover the `n == 0` case, but I think changing to that now would create more confusion than the increased cleanliness is worth.

[1]: https://github.com/opencontainers/image-spec/blob/656fb2f82121e8bd611b5aae4a6a3471e1bfe078/config.md#layer-chainid
[2]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc3/manifest.md#L45